### PR TITLE
zzf/add op bernoulli

### PR DESCRIPTION
**Pull request by Zhang-ze-feng**
## Motivation and Context
add op bernoulli

## Checklist
**Before PR**:

- [x] I have read and followed the workflow indicated in the [Contributors.md](https://github.com/DeepLink-org/DIOPI/blob/main/Contributors.md) to create this PR.
- [x] Pre-commit or linting tools indicated in [Contributors.md](https://github.com/DeepLink-org/DIOPI/blob/main/Contributors.md) are used to fix the potential lint issues.
- [x] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [x] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [x] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: impl/camb/functions/bernoulli.cpp [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+ * @copyright  (c) 2023, DeepLink.
+ */
+
+#include "../cnnl_helper.hpp"
+#include "../common/common.hpp"
+
+namespace impl {
+namespace camb {
+
+DIOPI_API diopiError_t diopiBernoulli(diopiContextHandle_t ctx, diopiTensorHandle_t out, diopiConstTensorHandle_t input, diopiGeneratorHandle_t generator) {
```
Comment (yangbofun) [inline impl/camb/functions/bernoulli.cpp:13]: generator，没有使用。

File: impl/camb/functions/bernoulli.cpp [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    cnnlHandle_t handle = cnnlHandlePool.get(ctx);
+    DiopiTensor inputTensor(input);
+
+    cnnlDataType_t dtype;
+    DIOPI_CALL(CnnlDataType::convertToCnnlType(&dtype, inputTensor.dtype()));
+    DiopiTensor outTemp = requiresTensor(ctx, inputTensor.shape(), inputTensor.dtype());
```
Comment (yangbofun) [inline impl/camb/functions/bernoulli.cpp:19]: 可以删除这句话，不用使用outTmp。直接使用out， 然后diopiLt那里使用diopiLtInp函数。

File: impl/camb/functions/bernoulli.cpp [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    return diopiSuccess;
+}
+
+DIOPI_API diopiError_t diopiBernoulliScalar(diopiContextHandle_t ctx, diopiTensorHandle_t out, double p, diopiGeneratorHandle_t generator) {
+    cnnlHandle_t handle = cnnlHandlePool.get(ctx);
+    DiopiTensor outTensor(out);
+
+    cnnlDataType_t dtype;
+    DIOPI_CALL(CnnlDataType::convertToCnnlType(&dtype, outTensor.dtype()));
+    DiopiTensor outTemp = requiresTensor(ctx, outTensor.shape(), outTensor.dtype());
```
Comment (yangbofun) [inline impl/camb/functions/bernoulli.cpp:63]: 同上，可以删除。下面的diopiLtScalar，可以调用diopiLtInpScalar.

File: impl/camb/functions/bernoulli.cpp [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    DIOPI_CALL(CnnlDataType::convertToCnnlType(&dtype, inputTensor.dtype()));
+
+    cnnlRandGenerator_t cnnlGenerator = nullptr;
+    DIOPI_CALLCNNL(cnnlRandCreateGenerator(&cnnlGenerator, CNNL_RAND_RNG_MTGP32));
+
+    diopiTensorHandle_t stateHandle = nullptr;
+    DIOPI_CALL(diopiGeneratorGetState(ctx, generator, &stateHandle));
+    void* statePtr = nullptr;
+    DIOPI_CALL(diopiGetTensorData(stateHandle, &statePtr));
+    DIOPI_CALLCNNL(cnnlRandGenerateUniform(handle, cnnlGenerator, dtype, statePtr, inputTensor.numel(), 0, 1, outputTensor.data()));
```
Comment (yangbofun) [inline impl/camb/functions/bernoulli.cpp:28]: inputTensor和outputTensor上的dtype的不一样的话，计算结果会出错。

File: impl/camb/functions/bernoulli.cpp [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    DIOPI_CALLCNNL(cnnlRandCreateGenerator(&cnnlGenerator, CNNL_RAND_RNG_MTGP32));
+
+    diopiTensorHandle_t stateHandle = nullptr;
+    DIOPI_CALL(diopiGeneratorGetState(ctx, generator, &stateHandle));
+    void* statePtr = nullptr;
+    DIOPI_CALL(diopiGetTensorData(stateHandle, &statePtr));
+    DIOPI_CALLCNNL(cnnlRandGenerateUniform(handle, cnnlGenerator, dtype, statePtr, inputTensor.numel(), 0, 1, outTemp.data()));
+    DIOPI_CALL(diopiLt(ctx, inout, diopiTensorHandle_t(outTemp), inout));
+
+    DIOPI_CALLCNNL(cnnlRandDestroyGenerator(cnnlGenerator));
```
Comment (caikun-pjlab) [inline impl/camb/functions/bernoulli.cpp:61]: 参考uniform算子的实现，调用完cnnlRandGenerateUniform算子之后，需要调用diopiGeneratorSetState将state写回到generator中

File: impl/camb/functions/bernoulli.cpp [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    DIOPI_CALLCNNL(cnnlRandCreateGenerator(&cnnlGenerator, CNNL_RAND_RNG_MTGP32));
+
+    diopiTensorHandle_t stateHandle = nullptr;
+    DIOPI_CALL(diopiGeneratorGetState(ctx, generator, &stateHandle));
+    void* statePtr = nullptr;
+    DIOPI_CALL(diopiGetTensorData(stateHandle, &statePtr));
+    DIOPI_CALLCNNL(cnnlRandGenerateUniform(handle, cnnlGenerator, dtype, statePtr, outTensor.numel(), 0, 1, outTensor.data()));
+
+    diopiScalar_t scalar = constructDiopiScalarT(outTensor.dtype(), p);
+    DIOPI_CALL(diopiLtInpScalar(ctx, out, &scalar));
```
Comment (caikun-pjlab) [inline impl/camb/functions/bernoulli.cpp:82]: 同上

**Review by caikun-pjlab — CHANGES_REQUESTED** [review]

**Review by caikun-pjlab — APPROVED** [review]

**Review by yangbofun — APPROVED** [review]

**Pull request merged**